File: Bulwark.Service/Reputation/RateLimiting.cs
namespace Bulwark.Service.Reputation;

/// <summary>
/// 简单令牌桶限流器。桶容量 = 每周期令牌数;以 (周期/容量) 的速率匀速补充。
/// <see cref="WaitAsync"/> 在无令牌时异步等待到下一个令牌可用,从而把请求速率
/// 平滑限制在配额内(如 4/min)。线程安全。
/// </summary>
public sealed class TokenBucket
{
    private readonly int _capacity;
    private readonly TimeSpan _refillInterval; // 每补充 1 个令牌的间隔
    private readonly SemaphoreSlim _mutex = new(1, 1);
    private double _tokens;
    private DateTime _lastRefillUtc;

    public TokenBucket(int tokensPerPeriod, TimeSpan period)
    {
        _capacity = Math.Max(1, tokensPerPeriod);
        _refillInterval = period / _capacity;
        _tokens = _capacity;      // 初始满桶
        _lastRefillUtc = DateTime.UtcNow;
    }

    /// <summary>取一个令牌;不足时异步等待到可用(受 token 取消约束)。</summary>
    public async Task WaitAsync(CancellationToken token = default)
    {
        while (true)
        {
            TimeSpan delay;
            await _mutex.WaitAsync(token);
            try
            {
                Refill();
                if (_tokens >= 1)
                {
                    _tokens -= 1;
                    return;
                }
                // 计算距离下一个令牌还需多久。
                var elapsed = DateTime.UtcNow - _lastRefillUtc;
                delay = _refillInterval - elapsed;
                if (delay < TimeSpan.Zero) delay = TimeSpan.Zero;
            }
            finally { _mutex.Release(); }

            await Task.Delay(delay <= TimeSpan.Zero ? TimeSpan.FromMilliseconds(50) : delay, token);
        }
    }

    private void Refill()
    {
        var now = DateTime.UtcNow;
        var elapsed = now - _lastRefillUtc;
        if (elapsed <= TimeSpan.Zero) return;

        var refill = elapsed.TotalMilliseconds / _refillInterval.TotalMilliseconds;
        if (refill >= 1)
        {
            _tokens = Math.Min(_capacity, _tokens + Math.Floor(refill));
            _lastRefillUtc = now;
        }
    }
}

/// <summary>
/// 每日配额计数器(本地)。跨自然日(UTC)自动重置。
/// 用于在发起请求前就拦掉超过日配额的查询,避免无谓的网络往返。线程安全。
/// </summary>
public sealed class DailyQuota
{
    private readonly int _limit;
    private readonly object _lock = new();
    private int _count;
    private DateTime _dayUtc;

    public DailyQuota(int dailyLimit)
    {
        _limit = Math.Max(1, dailyLimit);
        _dayUtc = DateTime.UtcNow.Date;
    }

    /// <summary>尝试消费一次配额。成功返回 true;当日已用尽返回 false。</summary>
    public bool TryConsume()
    {
        lock (_lock)
        {
            var today = DateTime.UtcNow.Date;
            if (today != _dayUtc)
            {
                _dayUtc = today;
                _count = 0;
            }
            if (_count >= _limit) return false;
            _count++;
            return true;
        }
    }
}

File: cpp/service/src/reputation/RateLimiting.cpp
#include "bulwark/service/reputation/RateLimiting.h"

#include <QThread>

#include <algorithm>
#include <cmath>

namespace bulwark::service::reputation {

TokenBucket::TokenBucket(int tokensPerPeriod, qint64 periodMs) {
    capacity_ = std::max(1, tokensPerPeriod);
    refillIntervalMs_ = static_cast<double>(periodMs) / capacity_;
    tokens_ = capacity_; // start full
    lastRefillUtc_ = QDateTime::currentDateTimeUtc();
}

void TokenBucket::refill() {
    const QDateTime now = QDateTime::currentDateTimeUtc();
    const qint64 elapsedMs = lastRefillUtc_.msecsTo(now);
    if (elapsedMs <= 0) return;
    const double refill = elapsedMs / refillIntervalMs_;
    if (refill >= 1.0) {
        tokens_ = std::min<double>(capacity_, tokens_ + std::floor(refill));
        lastRefillUtc_ = now;
    }
}

bool TokenBucket::tryConsume(bool priority) {
    QMutexLocker lk(&mutex_);
    refill();
    const bool yieldToPriority = !priority && priorityWaiters_.load() > 0;
    if (tokens_ >= 1.0 && !yieldToPriority) {
        tokens_ -= 1.0;
        return true;
    }
    return false;
}

void TokenBucket::wait(bool priority) {
    if (priority) priorityWaiters_.fetch_add(1);
    for (;;) {
        qint64 delayMs = 0;
        {
            QMutexLocker lk(&mutex_);
            refill();
            const bool yieldToPriority = !priority && priorityWaiters_.load() > 0;
            if (tokens_ >= 1.0 && !yieldToPriority) {
                tokens_ -= 1.0;
                if (priority) priorityWaiters_.fetch_sub(1);
                return;
            }
            const qint64 elapsed = lastRefillUtc_.msecsTo(QDateTime::currentDateTimeUtc());
            delayMs = static_cast<qint64>(refillIntervalMs_) - elapsed;
            if (delayMs < 0) delayMs = 0;
        }
        QThread::msleep(delayMs <= 0 ? 50 : static_cast<unsigned long>(delayMs));
    }
}

DailyQuota::DailyQuota(int dailyLimit, int priorityReserve) {
    limit_ = std::max(1, dailyLimit);
    priorityReserve_ = std::max(0, std::min(priorityReserve, limit_ - 1));
    dayUtc_ = QDateTime::currentDateTimeUtc().date();
}

bool DailyQuota::tryConsume(bool priority) {
    QMutexLocker lk(&lock_);
    const QDate today = QDateTime::currentDateTimeUtc().date();
    if (today != dayUtc_) { dayUtc_ = today; count_ = 0; }
    const int effectiveLimit = priority ? limit_ : (limit_ - priorityReserve_);
    if (count_ >= effectiveLimit) return false;
    ++count_;
    return true;
}

void DailyQuota::release() {
    QMutexLocker lk(&lock_);
    const QDate today = QDateTime::currentDateTimeUtc().date();
    if (today != dayUtc_) { dayUtc_ = today; count_ = 0; return; } // 跨天后计数已归零,无可退还
    if (count_ > 0) --count_;
}

std::pair<int, int> DailyQuota::snapshot() {
    QMutexLocker lk(&lock_);
    const QDate today = QDateTime::currentDateTimeUtc().date();
    if (today != dayUtc_) { dayUtc_ = today; count_ = 0; }
    return { count_, limit_ };
}

} // namespace bulwark::service::reputation
